### PR TITLE
Expose missing BinaryPrimitives APIs

### LINKED ISSUE #72107: [API Proposal] Expose missing BinaryPrimitives APIs
`System.Buffers.Binary.BinaryPrimitives` exposes many APIs but there are few notable ones missing for core number types. We should extend the type to ensure that it supports the basic operations for all basic number types.

```csharp
namespace System.Buffers.Binary;

public static partial class BinaryPrimitives
{
    // Read*BigEndian and Read*LittleEndian

    public static nint ReadIntPtrBigEndian(ReadOnlySpan<byte> source);
    public static nint ReadIntPtrLittleEndian(ReadOnlySpan<byte> source);

    public static Int128 ReadInt128BigEndian(ReadOnlySpan<byte> source);
    public static Int128 ReadInt128LittleEndian(ReadOnlySpan<byte> source);

    public static nuint ReadUIntPtrBigEndian(ReadOnlySpan<byte> source);
    public static nuint ReadUIntPtrLittleEndian(ReadOnlySpan<byte> source);

    public static UInt128 ReadUInt128BigEndian(ReadOnlySpan<byte> source);
    public static UInt128 ReadUInt128LittleEndian(ReadOnlySpan<byte> source);

    // ReverseEndianness

    public static nint ReverseEndianness(nint value);
    public static Int128 ReverseEndianness(Int128 value);
    public static nuint ReverseEndianness(nuint value);
    public static UInt128 ReverseEndianness(UInt128 value);

    // TryRead*BigEndian and TryRead*LittleEndian

    public static bool TryReadIntPtrBigEndian(ReadOnlySpan<byte> source, out nint value);
    public static bool TryReadIntPtrLittleEndian(ReadOnlySpan<byte> source, out nint value);

    public static bool TryReadInt128BigEndian(ReadOnlySpan<byte> source, out Int128 value);
    public static bool TryReadInt128LittleEndian(ReadOnlySpan<byte> source, out Int128 value);

    public static bool TryReadUIntPtrBigEndian(ReadOnlySpan<byte> source, out nuint value);
    public static bool TryReadUIntPtrLittleEndian(ReadOnlySpan<byte> source, out nuint value);

    public static bool TryReadUInt128BigEndian(ReadOnlySpan<byte> source, out UInt128 value);
    public static bool TryReadUInt128LittleEndian(ReadOnlySpan<byte> source, out UInt128 value);

    // TryWrite*BigEndian and TryWrite*LittleEndian

    public static bool TryWriteIntPtrBigEndian(ReadOnlySpan<byte> destination, nint value);
    public static bool TryWriteIntPtrLittleEndian(ReadOnlySpan<byte> destination, nint value);

    public static bool TryWriteInt128BigEndian(ReadOnlySpan<byte> destination, Int128 value);
    public static bool TryWriteInt128LittleEndian(ReadOnlySpan<byte> destination, Int128 value);

    public static bool TryWriteUIntPtrBigEndian(ReadOnlySpan<byte> destination, nuint value);
    public static bool TryWriteUIntPtrLittleEndian(ReadOnlySpan<byte> destination, nuint value);

    public static bool TryWriteUInt128BigEndian(ReadOnlySpan<byte> destination, UInt128 value);
    public static bool TryWriteUInt128LittleEndian(ReadOnlySpan<byte> destination, UInt128 value);

    // Write*BigEndian and Write*LittleEndian

    public static void WriteIntPtrBigEndian(ReadOnlySpan<byte> destination, nint value);
    public static void WriteIntPtrLittleEndian(ReadOnlySpan<byte> destination, nint value);

    public static void WriteInt128BigEndian(ReadOnlySpan<byte> destination, Int128 value);
    public static void WriteInt128LittleEndian(ReadOnlySpan<byte> destination, Int128 value);

    public static void WriteUIntPtrBigEndian(ReadOnlySpan<byte> destination, nuint value);
    public static void WriteUIntPtrLittleEndian(ReadOnlySpan<byte> destination, nuint value);

    public static void WriteUInt128BigEndian(ReadOnlySpan<byte> destination, UInt128 value);
    public static void WriteUInt128LittleEndian(ReadOnlySpan<byte> destination, UInt128 value);
}
```

### Additional Notes

Also missing are some APIs around `char`, `Half`, `float`, and `double`. However, there are concerns around `char` not being a "core" numeric type and around possible normalization issues for the various floating-point types, so they aren't included.

### PR DESCRIPTION
**Pull request by tannergooding**
This resolves #72107

### CODE REVIEW

**Comment by stephentoub** [comment]
Is there anywhere we should be using these?

**Comment by tannergooding** [comment]
I'm not aware of any places where we're using `Int128`/`UInt128` ourselves yet.

Looks like we can switch to using `IntPtr`/`UIntPtr` in `EventPipe`. I didn't transitively look at other calls (`BinaryReader.ReadInt64` or `XmlBinaryReader.GetInt64`), however.

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Binary/BinaryPrimitives.ReadBigEndian.cs [hunk 1]
```diff
@@ -75,6 +75,28 @@ public static long ReadInt64BigEndian(ReadOnlySpan<byte> source)
                 MemoryMarshal.Read<long>(source);
         }
 
+        /// <summary>
+        /// Reads an Int128 out of a read-only span of bytes as big endian.
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Buffers/Binary/BinaryPrimitives.ReadBigEndian.cs:79]: We have some variation in our XML comments here, but it looks like the official docs evolved for all of these to be more of the form:
```suggestion
        /// Reads a <see cref="Int128"/> from the beginning of a read-only span of bytes, as big endian.
```

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Buffers/Binary/BinaryPrimitives.ReadBigEndian.cs:79]: Did a general fixup for the APIs in the impacted files.

**Comment by tannergooding** [comment]
Fixed the docs and updated relevant places to use the new BinaryPrimitives APIs where applicable.

**Comment by tannergooding** [comment]
Failure is https://github.com/dotnet/runtime/issues/82055, this can be merged once it gets sign-off.

**Review by dakersnar — APPROVED** [review]
LGTM. Any specific logic in the tests that could use a deeper pass? Also, will we want these APIs on the new Decimal types?

**Comment by tannergooding** [comment]
> Any specific logic in the tests that could use a deeper pass?

None come to mind. These are relatively straightforward APIs.

> Also, will we want these APIs on the new Decimal types?

Probably not without explicit user ask. Users can achieve the same thing by reading the raw bits and calling `DecodeDecimal`/`DecodeBinary`.

**Pull request merged**